packages/flutter_tools/lib/src/web/devfs_web.dart

```diff
@@ -44,18 +47,49 @@ class WebAssetServer {
   ///
   /// Unhandled exceptions will throw a [ToolExit] with the error and stack
   /// trace.
-  static Future<WebAssetServer> start(String hostname, int port) async {
+  static Future<WebAssetServer> start(
+    String hostname,
+    int port,
+    UrlTunneller urlTunneller,
+    BuildMode buildMode,
+    bool enableDwds,
+  ) async {
     try {
       final InternetAddress address = (await InternetAddress.lookup(hostname)).first;
       final HttpServer httpServer = await HttpServer.bind(address, port);
       final Packages packages =
           await loadPackagesFile(Uri.base.resolve('.packages'), loader: (Uri uri) => globals.fs.file(uri).readAsBytes());
-      return WebAssetServer(httpServer, packages, address,
-          onError: (dynamic error, StackTrace stackTrace) {
-        httpServer.close(force: true);
-        throwToolExit(
-            'Unhandled exception in web development server:\n$error\n$stackTrace');
-      });
+      final WebAssetServer server = WebAssetServer(httpServer, packages, address);
+
+      // In debug builds, spin up DWDS and the full asset server.
+      if (buildMode == BuildMode.debug) {
+        final Dwds dwds = await Dwds.start(
+          assetReader: server,
+          buildResults: const Stream<BuildResult>.empty(),
+          chromeConnection: () async {
+            final Chrome chrome = await ChromeLauncher.connectedInstance;
+            return chrome.chromeConnection;
+          },
+          urlEncoder: urlTunneller,
+          enableDebugging: true,
+          logWriter: (Level logLevel, String message) => globals.printTrace(message)
+        );
+        shelf.Pipeline pipeline = const shelf.Pipeline();
+        if (enableDwds) {
+          pipeline = pipeline.addMiddleware(dwds.middleware);
+        }
+        final shelf.Handler dwdsHandler = pipeline.addHandler(server.handleRequest);
+        final shelf.Cascade cascade = shelf.Cascade()
+          .add(dwds.handler)
+          .add(dwdsHandler);
```

Comment:
Should this be behind `if (enableDwds)` too?

Reply:
No, we only need to disable the middleware, that is what hoists main and prevents the app from running